File: 01-Computer-Science/code/python/seb_cs/netdemo.py
from __future__ import annotations

import socket
import threading


def tcp_echo_once(message: str, host: str = "127.0.0.1") -> str:
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind((host, 0))
    server.listen(1)
    port = server.getsockname()[1]

    def serve() -> None:
        conn, _ = server.accept()
        with conn:
            data = conn.recv(4096)
            conn.sendall(data)

    thread = threading.Thread(target=serve, daemon=True)
    thread.start()
    with socket.create_connection((host, port), timeout=2) as client:
        client.sendall(message.encode("utf-8"))
        reply = client.recv(4096).decode("utf-8")
    server.close()
    thread.join(timeout=2)
    return reply


def udp_echo_once(message: str, host: str = "127.0.0.1") -> str:
    server = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    server.bind((host, 0))
    port = server.getsockname()[1]

    def serve() -> None:
        data, addr = server.recvfrom(4096)
        server.sendto(data, addr)

    thread = threading.Thread(target=serve, daemon=True)
    thread.start()
    client = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    client.settimeout(2)
    client.sendto(message.encode("utf-8"), (host, port))
    data, _ = client.recvfrom(4096)
    client.close()
    server.close()
    thread.join(timeout=2)
    return data.decode("utf-8")

File: 01-Computer-Science/code/typescript/src/netdemo.ts
import net from "node:net";
import dgram from "node:dgram";

export async function tcpEchoOnce(message: string, host = "127.0.0.1"): Promise<string> {
  const server = net.createServer((socket) => {
    socket.on("data", (data) => {
      socket.write(data);
      socket.end();
    });
  });
  await new Promise<void>((resolve) => server.listen(0, host, () => resolve()));
  const port = (server.address() as net.AddressInfo).port;
  try {
    const reply = await new Promise<string>((resolve, reject) => {
      const client = net.createConnection({ host, port }, () => {
        client.write(message);
      });
      let buf = "";
      client.on("data", (d) => {
        buf += d.toString("utf8");
      });
      client.on("end", () => resolve(buf));
      client.on("error", reject);
    });
    return reply;
  } finally {
    await new Promise<void>((resolve, reject) => server.close((err) => (err ? reject(err) : resolve())));
  }
}

export async function udpEchoOnce(message: string, host = "127.0.0.1"): Promise<string> {
  const server = dgram.createSocket("udp4");
  await new Promise<void>((resolve) =>
    server.bind(0, host, () => {
      server.on("message", (msg, rinfo) => {
        server.send(msg, rinfo.port, rinfo.address);
      });
      resolve();
    }),
  );
  const port = server.address().port;
  const client = dgram.createSocket("udp4");
  try {
    const reply = await new Promise<string>((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error("udp timeout")), 2000);
      client.on("message", (msg) => {
        clearTimeout(timer);
        resolve(msg.toString("utf8"));
      });
      client.send(message, port, host, (err) => {
        if (err) {
          clearTimeout(timer);
          reject(err);
        }
      });
    });
    return reply;
  } finally {
    client.close();
    server.close();
  }
}
